## Group 7
## Big Data Group Project

### Phase 2: Cleansing and Reshaping: Removing Duplicates & Columns with more than 30% null values 

#### Removing Columns with more than 30% null values

Insert libraries into the cell below:

In [1]:
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._
import org.apache.spark.sql._
import org.apache.hadoop.io.LongWritable
import org.apache.hadoop.io.Text
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.mapreduce.lib.input.TextInputFormat

Intitializing Scala interpreter ...

Spark Web UI available at http://95a19fa2a889:4040
SparkContext available as 'sc' (version = 2.4.3, master = local[*], app id = local-1558789912679)
SparkSession available as 'spark'


import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._
import org.apache.spark.sql._
import org.apache.hadoop.io.LongWritable
import org.apache.hadoop.io.Text
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.mapreduce.lib.input.TextInputFormat


Load the data into a scala dataframe:

In [1]:
val df = spark.read.format("com.databricks.spark.csv").option("header", "true").load("m_train_small.csv")

Intitializing Scala interpreter ...

Spark Web UI available at http://95a19fa2a889:4040
SparkContext available as 'sc' (version = 2.4.3, master = local[*], app id = local-1558848888025)
SparkSession available as 'spark'


df: org.apache.spark.sql.DataFrame = [EXT_SOURCE_1: string, EXT_SOURCE_2: string ... 359 more fields]


Count the rows in the dataframe:

In [7]:
df.count()

res3: Long = 307511


Check the percentage of NotNull value of each column:

In [6]:
import org.apache.spark.sql.functions.{col, count, when}

val df4=df.select(df.columns.map(c => (count(c) / count("*")).alias(c)): _*)

import org.apache.spark.sql.functions.{col, count, when}
df4: org.apache.spark.sql.DataFrame = [EXT_SOURCE_1: double, EXT_SOURCE_2: double ... 359 more fields]


In [7]:
df4.show()

+------------------+------------------+------------------+---------------------------------------+----------+----------+------------------+-------------+------------------------------+----------------------+---------------+----------------------------------+----------------------------+-------------------------------+-------------------+-------------------------------+-----------------------------------------+-----------------------------------+----------------------+-------------------------+--------------------------+-----------------+-------------------------------------------+-----------------------------------+-------------+--------------------------+-----------------------------------+-----------------------------+-------------------------+-------------------------------------------+---------------------------+---------------------------------------------------+-----------------------+----------------------------------------+-------------------------------------------+-----------

|      EXT_SOURCE_1|      EXT_SOURCE_2|      EXT_SOURCE_3|client_installments_AMT_PAYMENT_min_sum|DAYS_BIRTH|AMT_CREDIT|       AMT_ANNUITY|DAYS_EMPLOYED|bureau_DAYS_CREDIT_ENDDATE_max|bureau_DAYS_CREDIT_max|DAYS_ID_PUBLISH|bureau_AMT_CREDIT_MAX_OVERDUE_mean|bureau_DAYS_ENDDATE_FACT_max|bureau_AMT_CREDIT_SUM_DEBT_mean|        OWN_CAR_AGE|previous_loans_CNT_PAYMENT_mean|client_cash_CNT_INSTALMENT_FUTURE_min_max|previous_loans_SELLERPLACE_AREA_max|DAYS_LAST_PHONE_CHANGE|bureau_AMT_CREDIT_SUM_max|bureau_AMT_CREDIT_SUM_mean|DAYS_REGISTRATION|client_installments_DAYS_INSTALMENT_max_max|previous_loans_AMT_DOWN_PAYMENT_max|CODE_GENDER_F|REGION_POPULATION_RELATIVE|client_cash_CNT_INSTALMENT_max_mean|bureau_DAYS_CREDIT_UPDATE_max|bureau_AMT_CREDIT_SUM_sum|client_credit_AMT_DRAWINGS_CURRENT_mean_max|YEARS_BEGINEXPLUATATION_AVG|client_installments_NUM_INSTALMENT_VERSION_mean_max|bureau_DAYS_CREDIT_mean|client_installments_AMT_PAYMENT_sum_mean|client_installments_DAYS_INSTALMENT_min_max|previous_lo

TUAL_min_sum|previous_loans_HOUR_APPR_PROCESS_START_min|client_credit_AMT_INST_MIN_REGULARITY_max_min|client_cash_NAME_CONTRACT_STATUS_Signed_count_norm_max|OCCUPATION_TYPE_Laborers|client_cash_CNT_INSTALMENT_min_max|previous_loans_WEEKDAY_APPR_PROCESS_START_THURSDAY_count_norm|ORGANIZATION_TYPE_Construction|client_cash_NAME_CONTRACT_STATUS_Active_count_max|client_cash_SK_DPD_DEF_sum_min|previous_loans_NAME_CONTRACT_TYPE_Cash loans_count_norm|previous_loans_DAYS_FIRST_DRAWING_mean|client_credit_CNT_DRAWINGS_ATM_CURRENT_max_sum|previous_loans_PRODUCT_COMBINATION_Cash X-Sell: middle_count_norm|previous_loans_NAME_CASH_LOAN_PURPOSE_Other_count_norm|client_bureau_balance_STATUS_0_count_norm_sum|client_installments_NUM_INSTALMENT_VERSION_sum_min|previous_loans_CHANNEL_TYPE_Country-wide_count_norm|client_cash_NAME_CONTRACT_STATUS_Signed_count_sum|AMT_REQ_CREDIT_BUREAU_YEAR|bureau_CREDIT_TYPE_Microloan_count|NAME_INCOME_TYPE_State servant|client_credit_AMT_CREDIT_LIMIT_ACTUAL_sum_max|previous

|0.4361892745300168|0.9978537353135335|0.8017469293781361|                     0.9483985938714388|       1.0|       1.0|0.9999609770057006|          1.0|            0.8495370897301234|    0.8568506492450677|            1.0|                0.5979818608114832|          0.7490951543196829|             0.8329165460747746|0.34009189915157506|             0.9451434257636312|                       0.9411695841774765|                 0.9464929709831518|    0.9999967480838083|        0.856847397328876|         0.856847397328876|              1.0|                         0.9483985938714388|                  0.889740529607071|          1.0|                       1.0|                 0.9411695841774765|           0.8568506492450677|       0.8568506492450677|                         0.2826077766323806|         0.5121898078442723|                                 0.9483985938714388|     0.8568506492450677|                      0.9483985938714388|                         0.9483985938714388|           

 0.9464929709831518|                           0.2826077766323806|                          0.29992748226892696|                                     0.9464929709831518|                                    0.9464929709831518|                            0.2826077766323806|                         0.29992748226892696|             1.0|                                 0.9464929709831518|                      0.8568506492450677|                         0.2826077766323806|                                          0.2826077766323806|                     0.29992748226892696|                          0.9464929709831518|                                          0.9464929709831518|                        0.9464929709831518|                                   0.9464929709831518|                            1.0|                                                 0.9464929709831518|            0.9451434257636312|                        0.9464929709831518|                         1.0|                       

Check the count of null value in each column:

In [8]:
import org.apache.spark.sql.functions.{sum, col}
//df.select(df.columns.map(c => sum(col(c).isNull.cast("int")).alias(c)): _*).show
val df3=df.select(df.columns.map(c => sum(col(c).isNull.cast("int")).alias(c)): _*)

import org.apache.spark.sql.functions.{sum, col}
df3: org.apache.spark.sql.DataFrame = [EXT_SOURCE_1: bigint, EXT_SOURCE_2: bigint ... 359 more fields]


In [9]:
df3.show()

+------------+------------+------------+---------------------------------------+----------+----------+-----------+-------------+------------------------------+----------------------+---------------+----------------------------------+----------------------------+-------------------------------+-----------+-------------------------------+-----------------------------------------+-----------------------------------+----------------------+-------------------------+--------------------------+-----------------+-------------------------------------------+-----------------------------------+-------------+--------------------------+-----------------------------------+-----------------------------+-------------------------+-------------------------------------------+---------------------------+---------------------------------------------------+-----------------------+----------------------------------------+-------------------------------------------+--------------------------------------------

|EXT_SOURCE_1|EXT_SOURCE_2|EXT_SOURCE_3|client_installments_AMT_PAYMENT_min_sum|DAYS_BIRTH|AMT_CREDIT|AMT_ANNUITY|DAYS_EMPLOYED|bureau_DAYS_CREDIT_ENDDATE_max|bureau_DAYS_CREDIT_max|DAYS_ID_PUBLISH|bureau_AMT_CREDIT_MAX_OVERDUE_mean|bureau_DAYS_ENDDATE_FACT_max|bureau_AMT_CREDIT_SUM_DEBT_mean|OWN_CAR_AGE|previous_loans_CNT_PAYMENT_mean|client_cash_CNT_INSTALMENT_FUTURE_min_max|previous_loans_SELLERPLACE_AREA_max|DAYS_LAST_PHONE_CHANGE|bureau_AMT_CREDIT_SUM_max|bureau_AMT_CREDIT_SUM_mean|DAYS_REGISTRATION|client_installments_DAYS_INSTALMENT_max_max|previous_loans_AMT_DOWN_PAYMENT_max|CODE_GENDER_F|REGION_POPULATION_RELATIVE|client_cash_CNT_INSTALMENT_max_mean|bureau_DAYS_CREDIT_UPDATE_max|bureau_AMT_CREDIT_SUM_sum|client_credit_AMT_DRAWINGS_CURRENT_mean_max|YEARS_BEGINEXPLUATATION_AVG|client_installments_NUM_INSTALMENT_VERSION_mean_max|bureau_DAYS_CREDIT_mean|client_installments_AMT_PAYMENT_sum_mean|client_installments_DAYS_INSTALMENT_min_max|previous_loans_NAME_CONTRACT_STATUS_Refused_

_max|OCCUPATION_TYPE_Laborers|client_cash_CNT_INSTALMENT_min_max|previous_loans_WEEKDAY_APPR_PROCESS_START_THURSDAY_count_norm|ORGANIZATION_TYPE_Construction|client_cash_NAME_CONTRACT_STATUS_Active_count_max|client_cash_SK_DPD_DEF_sum_min|previous_loans_NAME_CONTRACT_TYPE_Cash loans_count_norm|previous_loans_DAYS_FIRST_DRAWING_mean|client_credit_CNT_DRAWINGS_ATM_CURRENT_max_sum|previous_loans_PRODUCT_COMBINATION_Cash X-Sell: middle_count_norm|previous_loans_NAME_CASH_LOAN_PURPOSE_Other_count_norm|client_bureau_balance_STATUS_0_count_norm_sum|client_installments_NUM_INSTALMENT_VERSION_sum_min|previous_loans_CHANNEL_TYPE_Country-wide_count_norm|client_cash_NAME_CONTRACT_STATUS_Signed_count_sum|AMT_REQ_CREDIT_BUREAU_YEAR|bureau_CREDIT_TYPE_Microloan_count|NAME_INCOME_TYPE_State servant|client_credit_AMT_CREDIT_LIMIT_ACTUAL_sum_max|previous_loans_DAYS_TERMINATION_max|client_credit_AMT_PAYMENT_TOTAL_CURRENT_max_max|FLAG_WORK_PHONE|previous_loans_NAME_GOODS_CATEGORY_Photo / Cinema Equipment_

|      173378|         660|       60965|                                  15868|         0|         0|         12|            0|                         46269|                 44020|              0|                            123625|                       77156|                          51380|     202929|                          16869|                                    18091|                              16454|                     1|                    44021|                     44021|                0|                                      15868|                              33906|            0|                         0|                              18091|                        44020|                    44020|                                     220606|                     150007|                                              15868|                  44020|                                   15868|                                      15868|                                            

              0|                                              16454|                                   44020|                                     220606|                                                      220606|                                  215280|                                       16454|                                                       16454|                                     16454|                                                16454|                              0|                                                              16454|                         16869|                                     16454|                           0|                                  0|                           220606|                                                  18067|                                            16454|                                                    16454|                                            15868|                                   220606|             

Or: Express the count of null value in each column in forms of array:

In [2]:
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types.IntegerType

import org.apache.spark.sql.functions._
import org.apache.spark.sql.types.IntegerType


In [4]:
val col= df.columns
var dfArray=col.map(colmn=>df.select(lit(colmn).as("colName"),sum(when(df(colmn).isNull,1).otherwise(0)).as("missingValues")))
dfArray.tail.foldLeft(dfArray.head)((acc,itr)=>acc.union(itr)).show

2019-05-25 11:42:17,072 WARN  [Thread-4] util.Utils (Logging.scala:logWarning(66)) - Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.debug.maxToStringFields' in SparkEnv.conf.
+--------------------+-------------+
|             colName|missingValues|
+--------------------+-------------+
|        EXT_SOURCE_1|       173378|
|        EXT_SOURCE_2|          660|
|        EXT_SOURCE_3|        60965|
|client_installmen...|        15868|
|          DAYS_BIRTH|            0|
|          AMT_CREDIT|            0|
|         AMT_ANNUITY|           12|
|       DAYS_EMPLOYED|            0|
|bureau_DAYS_CREDI...|        46269|
|bureau_DAYS_CREDI...|        44020|
|     DAYS_ID_PUBLISH|            0|
|bureau_AMT_CREDIT...|       123625|
|bureau_DAYS_ENDDA...|        77156|
|bureau_AMT_CREDIT...|        51380|
|         OWN_CAR_AGE|       202929|
|previous_loans_CN...|        16869|
|client_cash_CNT_I...|        18091|
|previous_loan

col: Array[String] = Array(EXT_SOURCE_1, EXT_SOURCE_2, EXT_SOURCE_3, client_installments_AMT_PAYMENT_min_sum, DAYS_BIRTH, AMT_CREDIT, AMT_ANNUITY, DAYS_EMPLOYED, bureau_DAYS_CREDIT_ENDDATE_max, bureau_DAYS_CREDIT_max, DAYS_ID_PUBLISH, bureau_AMT_CREDIT_MAX_OVERDUE_mean, bureau_DAYS_ENDDATE_FACT_max, bureau_AMT_CREDIT_SUM_DEBT_mean, OWN_CAR_AGE, previous_loans_CNT_PAYMENT_mean, client_cash_CNT_INSTALMENT_FUTURE_min_max, previous_loans_SELLERPLACE_AREA_max, DAYS_LAST_PHONE_CHANGE, bureau_AMT_CREDIT_SUM_max, bureau_AMT_CREDIT_SUM_mean, DAYS_REGISTRATION, client_installments_DAYS_INSTALMENT_max_max, previous_loans_AMT_DOWN_PAYMENT_max, CODE_GENDER_F, REGION_POPULATION_RELATIVE, client_cash_CNT_INSTALMENT_max_mean, bureau_DAYS_CREDIT_UPDATE_max, bureau_AMT_CREDIT_SUM_sum, client_credit_AMT_D...

From the above outcome we can see that there are lots of columns with a big portion of null value.
But if we remove all rows containing 'null', there would be only 800+ rows and 300+columns left, far from enought for
the following steps.

Hence we set a threshold as 92253, which is 30% of the number of total rows. And set a pipeline to select the columns 
in which the null counts is less than 92253, in other words, less than 30% of the column is emply.

In [ ]:
val null_thresh = 92253              

val to_keep = df.columns.filter(
    c => df.agg(
        sum(when(df(c).isNull, 1).otherwise(0)).alias(c)
    ).first().getLong(0) <= null_thresh
)

//df.select(to_keep.head, to_keep.tail: _*).show()

2019-05-26 05:35:52,874 WARN  [Thread-4] util.Utils (Logging.scala:logWarning(66)) - Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.debug.maxToStringFields' in SparkEnv.conf.


In [3]:
val null_thresh = 92253              

val to_keep = df.columns.filter(
    c => df.agg(
        sum(when(df(c).isNull, 1).otherwise(0)).alias(c)
    ).first().getLong(0) <= null_thresh
)

//df.select(to_keep.head, to_keep.tail: _*).show()

2019-05-25 13:16:07,393 WARN  [Thread-4] util.Utils (Logging.scala:logWarning(66)) - Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.debug.maxToStringFields' in SparkEnv.conf.


null_thresh: Int = 92253
to_keep: Array[String] = Array(EXT_SOURCE_2, EXT_SOURCE_3, client_installments_AMT_PAYMENT_min_sum, DAYS_BIRTH, AMT_CREDIT, AMT_ANNUITY, DAYS_EMPLOYED, bureau_DAYS_CREDIT_ENDDATE_max, bureau_DAYS_CREDIT_max, DAYS_ID_PUBLISH, bureau_DAYS_ENDDATE_FACT_max, bureau_AMT_CREDIT_SUM_DEBT_mean, previous_loans_CNT_PAYMENT_mean, client_cash_CNT_INSTALMENT_FUTURE_min_max, previous_loans_SELLERPLACE_AREA_max, DAYS_LAST_PHONE_CHANGE, bureau_AMT_CREDIT_SUM_max, bureau_AMT_CREDIT_SUM_mean, DAYS_REGISTRATION, client_installments_DAYS_INSTALMENT_max_max, previous_loans_AMT_DOWN_PAYMENT_max, CODE_GENDER_F, REGION_POPULATION_RELATIVE, client_cash_CNT_INSTALMENT_max_mean, bureau_DAYS_CREDIT_UPDATE_max, bureau_AMT_CREDIT_SUM_sum, client_installments_NUM_INSTALMENT_VERSION_mean_max, ...

In [6]:
val df5=df.select(to_keep.head, to_keep.tail: _*)

df5: org.apache.spark.sql.DataFrame = [EXT_SOURCE_2: string, EXT_SOURCE_3: string ... 283 more fields]


In [4]:
df5.show()

+-------------------+-------------------+---------------------------------------+----------+----------+-----------+-------------+------------------------------+----------------------+---------------+----------------------------+-------------------------------+-------------------------------+-----------------------------------------+-----------------------------------+----------------------+-------------------------+--------------------------+-----------------+-------------------------------------------+-----------------------------------+-------------+--------------------------+-----------------------------------+-----------------------------+-------------------------+---------------------------------------------------+-----------------------+----------------------------------------+-------------------------------------------+------------------------------------------------------+------------------------------------------+------------------------------------------+---------------------

|       EXT_SOURCE_2|       EXT_SOURCE_3|client_installments_AMT_PAYMENT_min_sum|DAYS_BIRTH|AMT_CREDIT|AMT_ANNUITY|DAYS_EMPLOYED|bureau_DAYS_CREDIT_ENDDATE_max|bureau_DAYS_CREDIT_max|DAYS_ID_PUBLISH|bureau_DAYS_ENDDATE_FACT_max|bureau_AMT_CREDIT_SUM_DEBT_mean|previous_loans_CNT_PAYMENT_mean|client_cash_CNT_INSTALMENT_FUTURE_min_max|previous_loans_SELLERPLACE_AREA_max|DAYS_LAST_PHONE_CHANGE|bureau_AMT_CREDIT_SUM_max|bureau_AMT_CREDIT_SUM_mean|DAYS_REGISTRATION|client_installments_DAYS_INSTALMENT_max_max|previous_loans_AMT_DOWN_PAYMENT_max|CODE_GENDER_F|REGION_POPULATION_RELATIVE|client_cash_CNT_INSTALMENT_max_mean|bureau_DAYS_CREDIT_UPDATE_max|bureau_AMT_CREDIT_SUM_sum|client_installments_NUM_INSTALMENT_VERSION_mean_max|bureau_DAYS_CREDIT_mean|client_installments_AMT_PAYMENT_sum_mean|client_installments_DAYS_INSTALMENT_min_max|previous_loans_NAME_CONTRACT_STATUS_Refused_count_norm|client_installments_AMT_INSTALMENT_max_max|client_cash_CNT_INSTALMENT_FUTURE_min_mean|client_installments_N

+-------------------+-------------------+---------------------------------------+----------+----------+-----------+-------------+------------------------------+----------------------+---------------+----------------------------+-------------------------------+-------------------------------+-----------------------------------------+-----------------------------------+----------------------+-------------------------+--------------------------+-----------------+-------------------------------------------+-----------------------------------+-------------+--------------------------+-----------------------------------+-----------------------------+-------------------------+---------------------------------------------------+-----------------------+----------------------------------------+-------------------------------------------+------------------------------------------------------+------------------------------------------+------------------------------------------+---------------------

----------------------------------------------+--------------------------------------------------+---------------------------------------------------------+----------------+---------------------------------------------------------+----------------------------------------------------------------+----------------------------------+-------------------------+--------------------------------------------------+------------------------------------+----------------------------------------------------+----------------------------------+-----------------------------------------------------------+--------------------------------------------------+--------------------------------------------------------+------------------------------------------------------+--------------------------------------------------------------------+---------------------------------------------+-----------------------------------------------------------------+---------------------------------------------------------------

| 0.6222457752555098|               null|                     1154108.2949999995|    -16765| 1293502.5|    35698.5|        -1188|                        1216.0|                -606.0|           -291|                      -540.0|                            0.0|                           10.0|                                      1.0|                             1400.0|                -828.0|                 810000.0|                254350.125|          -1186.0|                                     -536.0|                             6885.0|            1|      0.003540999999999999|                 10.285714285714286|                        -43.0|                1017400.5|                                 1.1428571428571428|               -1400.75|                       453952.2204000002|                                     -716.0|                                                   0.0|                                 560835.36|                       0.42857142857142855|                     

                                               0.0|                                                                  0.0|                               6.0|                                               0.0|                                               0.0|                                          0.0|                                               0.0|                                                0.3333333333333333|                                                    0.0|                                                   1.0|               0|                                                0.0|                                     0.0|                                         2.0|                                                         0.0|                                       2.0|                                                  0.0|                              0|                                                 0.3333333333333333|                           6.0|                       

| 0.6504416904014653|               null|                      994476.6900000005|    -19005|  312682.5|    29686.5|        -3039|                          null|                  null|          -2437|                        null|                           null|                           23.0|                                      3.0|                             8025.0|                -617.0|                     null|                      null|          -9833.0|                                      -11.0|                            66987.0|            1|                  0.008019|                 17.142857142857142|                         null|                     null|                                                2.0|                   null|                      232499.71968750012|                                     -151.0|                                    0.1111111111111111|                                 691786.89|                        1.4285714285714286|                     

                                         0.1111111111111111|                           0.0|                                       2.0|                           0|                                  1|                                                    0.0|                                              1.0|                                       0.1111111111111111|                                              1.0|                                                                        0.0|                                               0.0|                                  0|                                                                0.125|                       0|                                              0.0|                          null|                                        null|                                              2.0|                                   1|                                    0|                                       5.0|         0|     0|    100006|
| 0.3

| 0.3542247319929012| 0.6212263380626669|                      431757.3600000001|    -16941|  490495.5|    27517.5|        -1588|                         471.0|                 -78.0|           -477|                      -790.0|                        80019.0|                           14.0|                                      0.0|                              110.0|               -2536.0|                 267606.0|                  156148.5|          -4970.0|                                      -69.0|                            12145.5|            0|      0.035792000000000004|                 11.783132530120485|                        -16.0|                 468445.5|                                              1.125|     -757.3333333333335|                      236751.26657142857|                                     -339.0|                                                   0.0|                                432218.295|                                       0.0|                     

                                  0|                                       1.0|         1|     0|    100008|
| 0.7239998516953141| 0.4920600938649263|                      487600.7400000001|    -13778| 1560726.0|    41301.0|        -3130|                        1402.0|                -239.0|           -619|                      -313.0|              76953.53571428571|                            8.0|                                      9.0|                              300.0|               -1562.0|                1777500.0|                 266711.75|          -1213.0|                                      -43.0|                            22032.0|            1|      0.035792000000000004|                              7.875|                        -23.0|                4800811.5|                                                1.0|                -1271.5|                       75844.93235294118|                                      -43.0|                                                   0.

| 0.7142792864482229| 0.5406544504453575|                     273213.90000000014|    -18850| 1530000.0|    42075.0|         -449|                         689.0|               -1138.0|          -2379|                     -1138.0|                      174003.75|                           10.0|                                      0.0|                             8636.0|               -1070.0|                 675000.0|                  495000.0|          -4597.0|                                     -769.0|                                0.0|            0|                  0.003122|                               10.0|                        -18.0|                 990000.0|                                                1.0|                -1939.5|                               274492.08|                                    -1039.0|                                                   0.0|                                  27463.41|                                       0.0|                     

|0.20574728800732814| 0.7517237147741489|                      67537.53000000017|    -20099| 1019610.0|    33826.5|       365243|                        -860.0|               -1309.0|          -3514|                      -968.0|                            0.0|                           14.0|                                      0.0|                             1371.0|                   0.0|                 145242.0|                108807.075|          -7427.0|                                      -37.0|                            13594.5|            1|                  0.018634|                               19.2|                       -965.0|                 435228.3|                                                1.0|                -1773.0|                       499877.7528461542|                                    -1159.0|                                                  0.25|                                   46485.0|                                       0.0|                     

                                                   1.0|                         0|                                                                     0.25|                                              54.0|                                        2.000000000000004|               0|                                                      0.0|                                                             0.0|                                 0|                      0.0|                                               3.0|                                 0.0|                                                 0.0|                               0.0|                                                        0.0|                                               0.0|                                                     0.0|                                    0.9777777777777776|                                                                 0.0|                                          0.0|                    

| 0.6518623334244781|  0.363945238612397|                               37102.95|    -10197|  652500.0|    21177.0|         -679|                         723.0|                -376.0|           -738|                      -346.0|                       151642.8|                            9.0|                                      4.0|                             1500.0|                -844.0|                 900000.0|             341241.553125|          -4427.0|                                      -34.0|                                0.0|            1|                    0.0228|                             10.875|                        -25.0|              2729932.425|                                                1.0|              -1095.375|                      127137.04815789481|                                      -64.0|                                                   0.0|                                 14045.625|                                      0.75|                     

                                            0.0|                                          0.0|                                                              0.0|                                                                  0.5|                               6.0|                                               0.0|                                               0.0|                                          0.0|                                               0.0|                                                               0.0|                                                    0.5|                                                   0.0|               0|                                                0.0|                                     0.0|                                         1.0|                                                         0.0|                                       2.0|                                                  0.0|                              0|             

| 0.7150418188660659| 0.1766525794312139|                      284454.0000000001|    -13439|   80865.0|     5881.5|        -2717|                         845.0|                -128.0|          -3227|                      -347.0|                        12744.9|                           17.0|                                      1.0|                              200.0|               -2370.0|                  91264.5|         67854.85714285714|           -311.0|                                     -537.0|                            58050.0|            1|                  0.031329|                  19.34328358208955|                         -3.0|                 474984.0|                                 1.0344827586206895|     -618.4285714285714|                      203297.99837837805|                                    -1017.0|                                                   0.0|                                 102028.14|                        0.1791044776119403|                     

           4.0|                                                  0.0|                              0|                                                               0.75|                           8.0|                                       3.0|                           0|                                  1|                                                    0.0|                                              0.0|                                                      0.0|                                             10.0|                                                                        0.0|                                               1.0|                                  0|                                                                  0.0|                       0|                                              0.0|                           0.0|                                         0.0|                                             11.0|                                   0|       

| 0.6426562048311103|               null|                      627598.9800000002|    -14583|  773680.5|    32778.0|         -203|                          null|                  null|          -2056|                        null|                           null|                           10.0|                                      6.0|                             1200.0|                -188.0|                     null|                      null|           -615.0|                                      -29.0|                            57181.5|            1|                  0.010006|                               10.2|                         null|                     null|                                 1.7142857142857142|                   null|                            463834.80375|                                     -149.0|                                                   0.0|                                 94906.035|                                       1.2|                     

               null|                                              5.0|                                   0|                                    0|                                       0.0|         0|     0|    100018|
|   0.34663398139668| 0.6785676886853644|                               84148.92|     -8728|  299772.0|    20160.0|        -1157|                       10463.0|                -495.0|          -1368|                        null|                      122735.07|                           12.0|                                      0.0|                             1000.0|                -925.0|                 450000.0|                  360000.0|          -3494.0|                                     -714.0|                                0.0|            0|                  0.020713|                               12.0|                         -7.0|                 720000.0|                                                1.5|                 -495.0|                             

| 0.2363778398884225|0.06210303783729682|                              92765.385|    -12931|  509602.5|    26149.5|        -1317|                         277.0|                 -90.0|          -3866|                      -135.0|                      30772.125|                            7.0|                                      2.0|                              550.0|                  -3.0|                 135000.0|                  49871.25|          -6392.0|                                      -31.0|                                0.0|            0|                  0.018634|                                7.0|                        -28.0|                 199485.0|                                                1.0|                 -261.5|                       45906.73961538461|                                     -211.0|                                                   0.0|                                   8779.77|                                       1.0|                     

| 0.6835133461914255|               null|                     152101.25999999995|     -9776|  270000.0|    13500.0|         -191|                          null|                  null|          -2427|                        null|                           null|                            7.0|                                      5.0|                             6000.0|               -2811.0|                     null|                      null|          -4143.0|                                      -20.0|                             9922.5|            1|                  0.010966|                  9.333333333333334|                         null|                     null|                                 1.3333333333333333|                   null|                       55707.12782608695|                                     -230.0|                                                   0.0|                                  52236.18|                        1.4814814814814814|                     

                                           0.0|                                         0.3333333333333333|                                                     1.0|                         0|                                                       0.6666666666666666|                                              23.0|                                                      3.0|               0|                                                      0.5|                                                             0.0|                                 0|                     null|                                               6.0|                                null|                                 0.16666666666666666|                              null|                                                        0.0|                                               0.0|                                                     0.0|                                                   1.0|                          

| 0.5866171400119664| 0.4776491548517548|                               247330.8|    -11348|  544491.0|    17563.5|        -2038|                        1669.0|                -157.0|          -3964|                      -138.0|                       13703.85|                           12.5|                                      0.0|                             1500.0|               -1850.0|                 444078.0|        126591.71884615385|          -1021.0|                                     -142.0|                             9000.0|            1|                  0.015221|                 13.862068965517242|                         -5.0|              1645692.345|                                               1.25|    -1164.3846153846155|                      137660.64230769224|                                     -232.0|                                                   0.0|                                231819.525|                                       0.0|                     

                             0.0|                                    0.9333333333333332|                                                                 0.0|                                          0.0|                                                              0.0|                                                                  0.0|                               4.0|                                               0.0|                                               0.0|                                          0.0|                                               0.0|                                                               0.5|                                                    0.0|                                                   0.0|               0|                                                0.0|                                     0.0|                                         1.0|                                                        0.25|                                 

We can see that there are 283 columns left.

Now we can try to drop the null rows and see what we've got:

In [ ]:
val df6=df5.na.drop()
df6.count()

#### Remove duplicates rows:

In [ ]:
df6.dropDuplicates("SK_ID_CURR").count()

We can see that there are no duplicate rows in this dataframe

Now write our dataframe in a csv file:

In [ ]:
df6.write.option("header", "true").csv("m_train_small_remove_some_columns.csv")